File: sampleAPI.m
% Test script for Z3Score.com RESTful API
% For documentation pertaining to the API please visit
% https://github.com/amiyapatanaik/z3score-api
% This script demonstrates the basic functionalities of the Z3Score
% API. The API can be accessed from any language
% to know more see: https://en.wikipedia.org/wiki/Representational_state_transfer
% Patents pending (c)-2016 Amiya Patanaik amiyain@gmail.com
%
% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
% THE SOFTWARE.
%
% Requires cfslib MATLAB library: https://github.com/neurobittechnologies/cfslib-MATLAB
%

% Main settings
% https is NOT supported, use http
serverURL = 'http://z3score.com/api/v1';
email = 'email@domain.com';
key = 'API_KEY';

%Download cfslib-MATLAB from https://github.com/neurobittechnologies/cfslib-MATLAB
addpath('cfslib-MATLAB');
addpath('cfslib-MATLAB/utilities');
addpath('cfslib-MATLAB/utilities/jsonlab');
addpath('cfslib-MATLAB/utilities/encoder');
clc

%check license
try
    response = loadjson(urlreadpost([serverURL '/check'],...
                                        {'email',email,'key',key}));
catch
    disp('Server is unreachable');
    return
end

if response.status == 0,
    disp('License check failed');
    disp(['Error message: ' response.message])
    return
end

disp(response.message);
fprintf('API Call limit (hourly): %d, Epoch limit (daily): %d\n',response.call_limit, response.epoch_limit);

% Path to sample EDF file
filepath = [pwd '/test.edf'];

%Read the raw data
disp('Reading raw EDF data');
tic;
[header, signalHeader, signalCell] = blockEdfLoad(filepath);
t = toc;
fprintf('Time taken %.3f seconds\n',t);
%Number of channels
N = numel(signalHeader);
%Now select only C3, C4, EoG-l and EoG-r channels
disp('Here are all the channels:');
for k=1:N,
    fprintf('%d: %s\n',k,signalHeader(k).signal_labels);
end
C3N = input('Please select the C3:A2 channel number: ');
C4N = input('Please select the C4:A1 channel number: ');
ELN = input('Please select the EOGl:A2 channel number: ');
ERN = input('Please select the EOGr:A1 channel number: ');
EM = input('Please select the bipolar EMG channel number: ');

%Find out sampling rate
samplingRateEEG = signalHeader(C3N).samples_in_record/header.data_record_duration;
samplingRateEOG = signalHeader(ELN).samples_in_record/header.data_record_duration;
samplingRateEMG = signalHeader(EM).samples_in_record/header.data_record_duration;
num_epochs = floor(size(signalCell{C3N},1)/samplingRateEEG/30);

%Convert raw stream to a CFS and write to a file
disp('Converting EDF data to CFS and saving in test.cfs');
tic;
%Convert raw PSG stream to CFS stream
stream = streamCFS_V2(signalCell{C3N}, signalCell{C4N}, signalCell{ELN}, signalCell{ERN}, signalCell{EM}, samplingRateEEG, samplingRateEOG, samplingRateEMG);
fileID = fopen('test.cfs','w');
fwrite(fileID,stream,'*uint8','ieee-le');
fclose(fileID);
t = toc;
fprintf('Time taken %.3f seconds\n',t);

%Now ask server to score the cfs file generated earlier 
disp('Scoring CFS stream');
%You can also read the CFS stream from file like this:
%f = fopen('test.cfs'); stream = fread(f,Inf,'*uint8'); fclose(f);
%as we already have the stream we won't read the CFS file again
tic;
try
    response = loadjson(urlreadpost([serverURL '/score'], ... 
        {'email',email,'key',key,'file',stream}));
catch
    disp('Server is unreachable');
    return
end
t = toc;

if response.status == 0,
    disp('Error scoring data.');
    disp(['Error message:' response.message])
    return
end

fprintf('Time taken %.3f seconds.\nAPI calls left (hourly limits): %d, Epochs left (daily limits): %d \n',t, response.calls_left, response.epochs_left);
%Automatic sleep scores
scores = response.message;
%Save the sleep scores
csvwrite('test_score.csv',response.message)
%Read expert sleep scores
expert = csvread('test_expert.csv');
expert = expert(1:num_epochs);

C = confusionmat(scores(:,1),expert);
fprintf('Auto scoring agreement with expert scorer: %.2f%%\n',sum(scores(:,1) == expert)*100/num_epochs);
kappa(C);

disp('Done');

File: sampleAPI.py
#!/usr/bin/python
# Test script for Z3Score.com RESTful API
# For documentation pertaining to the API please visit
# https://github.com/amiyapatanaik/z3score-api
# This script demonstrates the basic functionalities of the Z3Score
# API. The API can be accessed from any language
# to know more see: https://en.wikipedia.org/wiki/Representational_state_transfer
# Patents pending (c)-2016 Amiya Patanaik amiyain@gmail.com
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
# requires pycfslib and pyedflib, install using pip: 
# pip install pycfslib 
# pip install pyedflib

from pycfslib import save_stream_v2
import pyedflib
import numpy as np
import csv
from requests import post
from time import time
from sklearn.metrics import cohen_kappa_score

serverURL = 'https://z3score.com/api/v1'
email = 'email@domain.com'
key = 'yourKey'

# Check license validity
response = post(serverURL+'/check',data={'email':email, 'key':key})
if response.status_code != 200:
    print("ERROR communicating with server")
    exit(0)
data = response.json()
if data['status'] == 0:
    print("License check failed")
    print(data['message'])
    exit(0)

print(data['message'])
print('API Call limit (hourly): %d, Epoch limit (daily): %d\n' % (data['call_limit'], data['epoch_limit']))

path = 'test.edf'
edf_file = pyedflib.EdfReader(path)
labels = edf_file.getSignalLabels()
samples = edf_file.getNSamples()

print("Here are the channel labels:")
for idx, label in enumerate(labels):
    print('%d. %s' % (idx+1, label))

C3 = int(input("Enter channel C3-A1 number: ")) - 1
C4 = int(input("Enter channel C4-A2 number: ")) - 1
EL = int(input("Enter channel EoGleft-A1 number: ")) - 1
ER = int(input("Enter channel EoGright-A2 number: ")) - 1
EM = int(input("Enter channel bipolar EMG number: ")) - 1

fsampling = [edf_file.getSampleFrequency(C3), edf_file.getSampleFrequency(EL),
                     edf_file.getSampleFrequency(EM)]


print("Reading EDF file...")
start_time = time()
EEG_C3 = np.asarray(edf_file.readSignal(C3))
EEG_C4 = np.asarray(edf_file.readSignal(C4))
EOGL = np.asarray(edf_file.readSignal(EL))
EOGR = np.asarray(edf_file.readSignal(ER))
EMG = np.asarray(edf_file.readSignal(EM))
elapsed_time = time() - start_time
print("Time taken: %.3f" % elapsed_time)


print("Converting to CFS and saving in test.cfs...")
start_time = time()
stream = save_stream_v2('test.cfs', EEG_C3, EEG_C4, EOGL, EOGR, EMG, fsampling)
elapsed_time = time() - start_time
print("Time taken: %.3f" % elapsed_time)

print('Now scoring')
start_time = time()
files = {'file': ('stream.cfs', stream)}
response = post(serverURL+'/score', files=files, data={'email':email, 'key':key})
elapsed_time = time() - start_time
print("Time taken: %.3f" % elapsed_time)

if response.status_code != 200:
    print("ERROR communicating with server")
    exit(0)

data = response.json()
if data['status'] == 0:
    print("Scoring failed\n")
    print(data['message'])
    exit(0)

scores = np.array(data['message'])
num_epochs = len(scores)

with open('test_expert.csv', 'rt') as f:
    Yb = sum([[int(x) for x in rec] for rec in csv.reader(f, delimiter=',')], [])

accuracy = sum(scores[:,0]==Yb[0:num_epochs])*100.0/num_epochs

kappa = cohen_kappa_score(scores[:,0], Yb[0:num_epochs])

print("Auto scoring agreement with expert scorer: %.2f%%, Kappa: %.3f" % (accuracy, kappa))
print("Done.")
